op-e2e/actions/fjord_fork_test.go

```diff
@@ -0,0 +1,145 @@
+	dp.DeployConfig.L2GenesisFjordTimeOffset = &fjordOffset
+	require.NoError(t, dp.DeployConfig.Check(), "must have valid config")
+
+	sd := e2eutils.Setup(t, dp, defaultAlloc)
+	log := testlog.Logger(t, log.LvlDebug)
+	_, _, _, sequencer, engine, verifier, _, _ := setupReorgTestActors(t, dp, sd, log)
+	ethCl := engine.EthClient()
+
+	// start op-nodes
+	sequencer.ActL2PipelineFull(t)
+	verifier.ActL2PipelineFull(t)
+
+	// Get gas price from oracle
+	gasPriceOracle, err := bindings.NewGasPriceOracleCaller(predeploys.GasPriceOracleAddr, ethCl)
+	require.NoError(t, err)
+
+	// Get current implementations addresses (by slot) for L1Block + GasPriceOracle
+	initialGasPriceOracleAddress, err := ethCl.StorageAt(context.Background(), predeploys.GasPriceOracleAddr, genesis.ImplementationSlot, nil)
+	require.NoError(t, err)
+
+	sequencer.ActBuildL2ToFjord(t)
+
+	// get latest block
+	latestBlock, err := ethCl.BlockByNumber(context.Background(), nil)
+	require.NoError(t, err)
+	require.Equal(t, sequencer.L2Unsafe().Number, latestBlock.Number().Uint64())
+
+	transactions := latestBlock.Transactions()
+	// L1Block: 1 set-L1-info + 1 deploys + 1 upgradeTo + 1 enable fjord on GPO
+	// See [derive.FjordNetworkUpgradeTransactions]
+	require.Equal(t, 4, len(transactions))
+
+	// All transactions are successful
+	for i := 1; i < 4; i++ {
+		txn := transactions[i]
+		receipt, err := ethCl.TransactionReceipt(context.Background(), txn.Hash())
+		require.NoError(t, err)
+		require.Equal(t, types.ReceiptStatusSuccessful, receipt.Status)
+		require.NotEmpty(t, txn.Data(), "upgrade tx must provide input data")
+	}
+
+	expectedGasPriceOracleAddress := crypto.CreateAddress(derive.GasPriceOracleFjordDeployerAddress, 0)
+
+	// Gas Price Oracle Proxy is updated
+	updatedGasPriceOracleAddress, err := ethCl.StorageAt(context.Background(), predeploys.GasPriceOracleAddr, genesis.ImplementationSlot, latestBlock.Number())
+	require.NoError(t, err)
+	require.Equal(t, expectedGasPriceOracleAddress, common.BytesToAddress(updatedGasPriceOracleAddress))
+	require.NotEqualf(t, initialGasPriceOracleAddress, updatedGasPriceOracleAddress, "Gas Price Oracle Proxy address should have changed")
+	verifyCodeHashMatches(t, ethCl, expectedGasPriceOracleAddress, fjordGasPriceOracleCodeHash)
+
+	// Check that Fjord was activated
+	isFjord, err := gasPriceOracle.IsFjord(nil)
+	require.NoError(t, err)
+	require.True(t, isFjord)
+
+	// Check GetL1GasUsed is updated
+	txData, err := hex.DecodeString(txHex)
+	require.NoError(t, err)
+
+	gpoL1GasUsed, err := gasPriceOracle.GetL1GasUsed(&bind.CallOpts{}, txData)
+	require.NoError(t, err)
+	require.Equal(gt, uint64(1_888), gpoL1GasUsed.Uint64())
+
+	// Check that GetL1Fee takes into account fast LZ
+	gpoFee, err := gasPriceOracle.GetL1Fee(&bind.CallOpts{}, txData)
+	require.NoError(t, err)
+
+	gethFee := fjordL1Cost(t, gasPriceOracle, types.RollupCostData{
+		FastLzSize: uint64(types.FlzCompressLen(txData) + 68),
```

Comment:
I don't love this 68 constant floating around, seems like there should be an abstraction that "just works" ie `FastLzSize: FastLzFjord(txData)` that is used everywhere that encapsulates the type casting and addition

Reply:
Altho i have not looked at the geth implementation, so I may have no idea what I am talking about

Reply:
@danyalprout Actually a good point. Is the tx in `txHex` without signature? Because if it is with signature, then the 68 should be removed here and in line 120. Because it's added in both places, it doesn't change the passing of the tests.

Reply:
Regardless of whether the transaction is signed, I believe we still have to add the additional 68 in both of these cases. There are two separate checks here:

* GetL1Fee (lines 107-114)
* GetL1FeeUpperBound (lines 117-124)

The intention is to check that the output of the GasPriceOracle functions matches that of the Fjord Geth cost function for both of these cases.

For these functions the GPO takes a `[]byte` and will always add 68 to the length (regardless of whether the underlying txn where the bytes came from are signed or not). Whereas the cost function takes a RollupCostData struct with a FastLz property set, and does not add 68 as the assumption is it includes signature. To make the outputs match, we have to ensure that the data sizes passed through to the linear regression are the same. I think we can either do this via:

1. Explicitly adding the signature length to what we pass to the cost function (i.e. mirror the GPO behavior)
2. Removing the signature bytes from what we pass to the GPO 

I went with 1), but 2) maybe more obvious. Happy to update if you think it is.

Reply:
OK I now realize that this part of the test compares `gpoFee` (line 108) with `gethFee` (line 111), so it does make sense that both inputs differ by 68, because the geth implementation assumes to receive an encoded tx _including_ the sig, while the GPO assumes to receive a tx _excluding_ the sig.

What confused me was that the `txData` includes the signature, so 2) would have been more intuitive to me.
```
cast dt 02f875822105819b8405709fb884057d460082e97f94273ca93a52b817294830ed7572aa591ccfa647fd80881249c58b0021fb3fc080a05bb08ccfd68f83392e446dac64d88a2d28e7072c06502dfabc4a77e77b5c7913a05878d53dd4ebba4f6367e572d524dffcabeec3abb1d8725ee3ac5dc32e1852e3
{
  "type": "0x2",
  "chainId": "0x2105",
  "nonce": "0x9b",
  "gasLimit": "0xe97f",
  "maxFeePerGas": "0x57d4600",
  "maxPriorityFeePerGas": "0x5709fb8",
  "to": "0x273ca93a52b817294830ed7572aa591ccfa647fd",
  "value": "0x0",
  "accessList": [],
  "input": "0x1249c58b0021fb3f",
  "r": "0x5bb08ccfd68f83392e446dac64d88a2d28e7072c06502dfabc4a77e77b5c7913",
  "s": "0x5878d53dd4ebba4f6367e572d524dffcabeec3abb1d8725ee3ac5dc32e1852e3",
  "yParity": "0x0",
  "hash": "0x8debb2fe54200183fb8baa3c6dbd8e6ec2e4f7a4add87416cd60336b8326d16a"
}
```
But since both cost functions don't care whether the passed data is any garbage or an actual encoded tx, with or without signature, the test works. 1) would have made more sense to me if it had been an rlp-encoded _unsigned_ tx. Up to you if you want to change the test at all, to either do 2) or 1) with a stripped tx encoding.

Reply:
Lets leave this as is for now -- I think doing the +68 is more reliable than stripping the signature. (incase the signature can compress to less than 68 bytes).